File: src/imagebuilder/flyingtux/buildjail/docker.py
from metux.util import log, fs
from metux.util.docker import Docker
from metux.util.specobject import SpecObject
from os.path import dirname
from os import remove, rmdir

class BuildJailDriverDocker(SpecObject):

    rootfs_tarball = '/sysroot.tar.bz2'
    build_sysroot  = '/sysroot'
    spec_mandatory = [ 'workdir', 'rootfs-image', 'conf' ]

    def post_init(self):
        self.check_mandatory(self.spec_mandatory)
        self.my_docker = Docker()

    def __temp_fn(self, fn):
        return "%s/sysroot/%s" % (self['workdir'], fn)

    def sysroot_fn(self, fn):
        return "%s/%s" % (self.build_sysroot, fn)

    def sysroot_create_dirs(self, dirs):
        self.my_container.mkdirs([self.sysroot_fn(x) for x in dirs])
        for x in dirs:
            fs.mkdir(self.__temp_fn(x))

    def sysroot_write_file(self, fn, text):
        local = self.__temp_fn(fn)
        remote = self.sysroot_fn(fn)
        open(local, 'w').write(text)
        self.my_container.cp_to(local, remote)
        self.my_container.chown('root:root', [remote])

    def sysroot_copy_file(self, src, dst, mode=None, owner=None):
        self.my_container.cp_to(src, self.sysroot_fn(dst), mode=mode, owner=owner)

    def sysroot_remove_recursive(self, dirs):
        return self.my_container.rm_recursive([self.sysroot_fn(x) for x in dirs])

    def sysroot_remove_dirs(self, dirs):
        return self.my_container.exec_catch(
            ['find'] +
            [self.sysroot_fn(x) for x in dirs] +
            ['-type', 'd', '-empty', '-exec', 'rmdir', '-p', '{}', ';'])

    def sysroot_symlink(self, name, target):
        self.sysroot_create_dirs([dirname(name), dirname(target)])
        self.my_container.execute(['ln', '-s', target, self.sysroot_fn(name)])

    def call_cmd(self, args):
        return self.my_container.execute(args)

    def prepare(self):
        self.my_container = self.my_docker.container_create(self['conf::image'], [self['conf::init']])
        self.my_container.start()

    def finish(self):
        target_tarball = self['workdir']+'/rootfs-temp.tar.bz2'
        target_dir     = dirname(target_tarball)

        self.my_container.execute(['tar','-cf', self.rootfs_tarball, '-C', self.build_sysroot, '.'])
        fs.mkdir(target_dir)
        self.my_container.cp_from(self.rootfs_tarball, target_tarball)
        self.my_container.destroy()
        self.my_docker.image_import(target_tarball, self['rootfs-image'])

File: internal/buildjail/docker.go
// SPDX-License-Identifier: AGPL-3.0-or-later

package buildjail

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/metux/flyingtux/internal/dockercli"
	"github.com/metux/flyingtux/internal/fs"
	"github.com/metux/flyingtux/internal/util"
)

const (
	rootfsTarball = "/sysroot.tar.bz2"
	buildSysroot  = "/sysroot"
)

// dockerJail is the docker-backed BuildJail (port of BuildJailDriverDocker).
type dockerJail struct {
	workdir     string
	rootfsImage string
	confImage   string
	confInit    string
	docker      *dockercli.Docker
	container   *dockercli.Container
}

func newDocker(p Params) (BuildJail, error) {
	if p.Workdir == "" || p.RootfsImage == "" || p.ConfImage == "" {
		return nil, util.ConfigError("build jail: missing mandatory field (workdir/rootfs-image/conf)")
	}
	return &dockerJail{
		workdir:     p.Workdir,
		rootfsImage: p.RootfsImage,
		confImage:   p.ConfImage,
		confInit:    p.ConfInit,
		docker:      dockercli.New(),
	}, nil
}

func (j *dockerJail) tempFn(fn string) string {
	return fmt.Sprintf("%s/sysroot/%s", j.workdir, fn)
}

func (j *dockerJail) SysrootFn(fn string) string {
	return fmt.Sprintf("%s/%s", buildSysroot, fn)
}

func (j *dockerJail) SysrootCreateDirs(dirs []string) error {
	remote := make([]string, 0, len(dirs))
	for _, x := range dirs {
		remote = append(remote, j.SysrootFn(x))
	}
	j.container.Mkdirs(remote)
	for _, x := range dirs {
		if err := fs.Mkdir(j.tempFn(x)); err != nil {
			return err
		}
	}
	return nil
}

func (j *dockerJail) SysrootWriteFile(fn, text string) error {
	local := j.tempFn(fn)
	remote := j.SysrootFn(fn)
	if err := os.MkdirAll(filepath.Dir(local), 0o755); err != nil {
		return err
	}
	if err := os.WriteFile(local, []byte(text), 0o644); err != nil {
		return err
	}
	j.container.CpTo(local, remote, "", "")
	j.container.Chown("root:root", []string{remote})
	return nil
}

func (j *dockerJail) SysrootCopyFile(src, dst, mode, owner string) error {
	j.container.CpTo(src, j.SysrootFn(dst), owner, mode)
	return nil
}

func (j *dockerJail) SysrootRemoveRecursive(dirs []string) error {
	remote := make([]string, 0, len(dirs))
	for _, x := range dirs {
		remote = append(remote, j.SysrootFn(x))
	}
	j.container.RmRecursive(remote)
	return nil
}

func (j *dockerJail) SysrootRemoveDirs(dirs []string) error {
	args := []string{"find"}
	for _, x := range dirs {
		args = append(args, j.SysrootFn(x))
	}
	args = append(args, "-type", "d", "-empty", "-exec", "rmdir", "-p", "{}", ";")
	j.container.ExecCatch(args)
	return nil
}

func (j *dockerJail) SysrootSymlink(name, target string) error {
	if err := j.SysrootCreateDirs([]string{filepath.Dir(name), filepath.Dir(target)}); err != nil {
		return err
	}
	j.container.Execute([]string{"ln", "-s", target, j.SysrootFn(name)})
	return nil
}

func (j *dockerJail) CallCmd(args []string) int {
	return j.container.Execute(args)
}

func (j *dockerJail) Prepare() error {
	c, err := j.docker.ContainerCreate(j.confImage, []string{j.confInit}, "", nil, nil, false)
	if err != nil {
		return err
	}
	j.container = c
	j.container.Start()
	return nil
}

func (j *dockerJail) Finish() error {
	targetTarball := j.workdir + "/rootfs-temp.tar.bz2"
	targetDir := filepath.Dir(targetTarball)

	j.container.Execute([]string{"tar", "-cf", rootfsTarball, "-C", buildSysroot, "."})
	if err := fs.Mkdir(targetDir); err != nil {
		return err
	}
	j.container.CpFrom(rootfsTarball, targetTarball)
	j.container.Destroy()
	if _, err := j.docker.ImageImport(targetTarball, j.rootfsImage); err != nil {
		return err
	}
	return nil
}
